File: templates/partition.java
// template 1 : recommend
public int partitionArray(int[] nums, int start, int end, int k) {
    if(nums == null || nums.length == 0){
        return 0;
    }

    int pivot = k;
    int i = start;
    for (int j = i; j <= end; j++) {
        if (nums[j] < pivot) {
            swap(nums, i, j);
            i = i + 1;
        }
    }

    return i; // if all the num smaller than pivot, i would be nums.length, which can cause indexOutofRange in some cases
}


// template 2: two pointers
public int partition(int[] nums, int start, int end) {
    int left = start, right = end;
    int pivot = nums[left];

    while (left < right) {
        while (left < right && nums[right] >= pivot) {
            right--;
        }

        nums[left] = nums[right];

        while (left < right && nums[left] <= pivot) {
            left++;
        }

        nums[right] = nums[left];
    }

    nums[left] = pivot;
    // left and right points to same position
    return left;
}

File: partition.py
# 这个模版需要将pivot点设置在最左边
def partition(nums, start, end):
  left, right = start, end
  pivot = nums[left]
  while left < right:
    while left < right and nums[right] >= pivot:
      right -= 1
    nums[left] = nums[right]
    while left < right and nums[left] <= pivot:
      left += 1
    nums[right] = nums[left]

  nums[left] = pivot
  return left    

def partition(nums, k):
  left, right = start, end
  while left < right:
    while left < right and nums[left] < k:
      left += 1

    while left < right and nums[right] >= k:
      right -= 1

    if left < right:    
      nums[left], nums[right] = nums[right], nums[left]
      left += 1
      right -= 1

  return left    